lib/doctest.js

```diff
@@ -452,47 +418,35 @@ const run = queue =>
     return [{lines: {input: input.lines, output: output.lines}, comparison}];
   });
 
-const evaluateModule = path => source => {
-  const abspath = resolve (path)
-                  .replace (/[.][^.]+$/, `-${Date.now ()}.mjs`);
-
-  const cleanup = f => x =>
-    fs.promises.unlink (abspath)
-    .then (() => f (x));
-
-  return fs.promises.writeFile (abspath, source)
-    .then (() => import (abspath))
-    .then (module => run (module.__doctest.queue))
-    .then (cleanup (Promise.resolve.bind (Promise)),
-           cleanup (Promise.reject.bind (Promise)));
-};
-
-const commonjsEval = path => async source => {
-  const abspath = resolve (path)
-                  .replace (/[.][^.]+$/, `-${Date.now ()}.cjs`);
-
-  fs.writeFileSync (abspath, source);
-  let queue;
-  try {
-    ({queue} = (require (abspath)).__doctest);
-  } finally {
-    fs.unlinkSync (abspath);
-  }
+const evaluateModule = async source => {
+  const queue = [];
+  const enqueue = io => { queue.push (io); };
+  const __doctest = {enqueue};
+  const context = vm.createContext ({...global, __doctest});
+  const module = new vm.SourceTextModule (source, {context});
+  await module.link (async (specifier, referencingModule) => {
+    const entries = Object.entries (await import (specifier));
+    const module = new vm.SyntheticModule (
+      entries.map (([name]) => name),
+      () => {
+        for (const [name, value] of entries) {
+          module.setExport (name, value);
+        }
+      },
+      {identifier: specifier, context: referencingModule.context}
+    );
+    return module;
+  });
+  await module.evaluate ();
```

Comment:
I've never seen this vm API used! Can you explain briefly what's going on here?

Reply:
It's not clear to me. My understanding is that we're responsible for handling uses of `import` and `import()` within the module we're evaluating in memory. This must be necessary to handle relative import paths.

I used a modified version of the snippet in a [Stack Overflow answer][1].


[1]: https://stackoverflow.com/a/73282303/312785